### PR TITLE
Numerically Stable Decoders

### LINKED ISSUE #176: Consider making all link decoders output logits for numerical stability

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
### Summary / Description

Make all our decoders output logits for numerically stable binary cross entropy loss.

**Note**: Our TGN convergence is **much** better. Previously, we had a bug where we were doing `sigmoid()` twice, effectively. 

Close #176 

| Model   |  MSE / master |  BCE / dev |
|---------|----------------|----------------|
| GCLSTM (10 epochs)  |       0.2283 Test MRR        |      0.2598   Test MRR       |
| GCN  (10 epochs)   |         0.4072 Test MRR      |     0.4586 Test MRR          |
| TGN  (5 epochs)   |         0.0501 Test MRR      |     0.3441 Test MRR          |

### CODE REVIEW

File: examples/linkproppred/gcn.py [hunk 1]
```diff
@@ -117,8 +117,8 @@ def train(
         pos_out = decoder(z[batch.src], z[batch.dst])
         neg_out = decoder(z[batch.src], z[batch.neg])
 
-        loss = F.mse_loss(pos_out, torch.ones_like(pos_out))
-        loss += F.mse_loss(neg_out, torch.zeros_like(neg_out))
+        loss = F.binary_cross_entopy_with_logits(pos_out, torch.ones_like(pos_out))
```
Comment (Jacob-Chmura) [inline examples/linkproppred/gcn.py]: Why were we using MSE loss?

File: examples/linkproppred/gclstm.py [hunk 2]
```diff
@@ -96,8 +96,8 @@ def train(
         pos_out = decoder(z[batch.src], z[batch.dst])
         neg_out = decoder(z[batch.src], z[batch.neg])
 
-        loss = F.mse_loss(pos_out, torch.ones_like(pos_out))
-        loss += F.mse_loss(neg_out, torch.zeros_like(neg_out))
+        loss = F.binary_cross_entopy_with_logits(pos_out, torch.ones_like(pos_out))
```
Comment (Jacob-Chmura) [inline examples/linkproppred/gclstm.py]: Why were we using MSe loss?

File: examples/linkproppred/tgn.py
Comment (Jacob-Chmura) [inline examples/linkproppred/tgn.py:1]: TGN was already doing `F.binary_cross_entropy_with_logits`

File: examples/linkproppred/gclstm.py [hunk 3]
```diff
@@ -96,8 +96,8 @@ def train(
         pos_out = decoder(z[batch.src], z[batch.dst])
         neg_out = decoder(z[batch.src], z[batch.neg])
 
-        loss = F.mse_loss(pos_out, torch.ones_like(pos_out))
-        loss += F.mse_loss(neg_out, torch.zeros_like(neg_out))
+        loss = F.binary_cross_entropy_with_logits(pos_out, torch.ones_like(pos_out))
```
Comment (shenyangHuang) [inline examples/linkproppred/gclstm.py:99]: if we change this, we should run to make sure we can get the same performance as before

File: examples/linkproppred/gcn.py [hunk 4]
```diff
@@ -117,8 +117,8 @@ def train(
         pos_out = decoder(z[batch.src], z[batch.dst])
         neg_out = decoder(z[batch.src], z[batch.neg])
 
-        loss = F.mse_loss(pos_out, torch.ones_like(pos_out))
-        loss += F.mse_loss(neg_out, torch.zeros_like(neg_out))
+        loss = F.binary_cross_entropy_with_logits(pos_out, torch.ones_like(pos_out))
```
Comment (shenyangHuang) [inline examples/linkproppred/gcn.py:120]: similar here, ensure same performance before changing this, I use this loss in UTG for DTDG methods.

**Review by shenyangHuang — DISMISSED** [review]

File: examples/linkproppred/tgn.py
Comment (ntgbaoo) [reply examples/linkproppred/tgn.py:1]: So we are doing `.sigmoid()` twice in master right now? How is the performance of `TGN` on this branch compared to `main`? (just curious)

**Review by benjaminnNgo — CHANGES_REQUESTED** [review]
I am not sure if we need to call `.sigmoid()` after performing `encoder` forward pass in every `eval()` function? Since we now remove `.sigmoid()` in decoders and let criterion takes care of it in `train`, we no longer have `.sigmoid()` in `eval()`?

**Comment by Jacob-Chmura** [comment]
> I am not sure if we need to call `.sigmoid()` after performing `encoder` forward pass in every `eval()` function?

You are completely right, thanks for catching

File: examples/linkproppred/tgn.py
Comment (Jacob-Chmura) [reply examples/linkproppred/tgn.py:1]: Effectively, yes. I am double checking perf

**Review by benjaminnNgo — APPROVED** [review]
Looks good to me.
